mesonbuild/compilers.py

```diff
@@ -998,44 +998,75 @@ def run(self, code, env, extra_args=None, dependencies=None):
-                    return i - 1
-                return i
+    def _compile_int(self, expression, prefix, env, extra_args, dependencies):
+        fargs = {'prefix': prefix, 'expression': expression}
+        t = '''#include <stdio.h>
+        {prefix}
+        int main() {{ static int a[1-2*!({expression})]; a[0]=0; return 0; }}'''
+        return self.compiles(t.format(**fargs), env, extra_args, dependencies)
+
+    def cross_compute_int(self, expression, l, h, guess, prefix, env, extra_args, dependencies):
+        if isinstance(guess, int):
+            if self._compile_int('%s == %d' % (expression, guess), prefix, env, extra_args, dependencies):
+                return guess
+
+        cur = l
+        while l < h:
+            cur = int((l + h) / 2)
+            if cur == l:
+                break
+
+            if self._compile_int('%s >= %d' % (expression, cur), prefix, env, extra_args, dependencies):
+                l = cur
+            else:
+                h = cur
+
+        if self._compile_int('%s == %d' % (expression, cur), prefix, env, extra_args, dependencies):
+            return cur
         raise EnvironmentException('Cross-compile check overflowed')
 
-    def cross_sizeof(self, element, prefix, env, extra_args=None, dependencies=None):
+    def compute_int(self, expression, l, h, guess, prefix, env, extra_args=None, dependencies=None):
+        if extra_args is None:
+            extra_args = []
+        if self.is_cross:
+            return self.cross_compute_int(expression, l, h, guess, prefix, env, extra_args, dependencies)
+        fargs = {'prefix': prefix, 'expression': expression}
+        t = '''#include<stdio.h>
+        {prefix}
+        int main(int argc, char **argv) {{
+            printf("%ld\\n", (long)({expression}));
+            return 0;
+        }};'''
+        res = self.run(t.format(**fargs), env, extra_args, dependencies)
+        if not res.compiled:
+            return -1
+        if res.returncode != 0:
+            raise EnvironmentException('Could not run compute_int test binary.')
+        return int(res.stdout)
+
+    def cross_sizeof(self, typename, prefix, env, extra_args=None, dependencies=None):
         if extra_args is None:
             extra_args = []
-        fargs = {'prefix': prefix, 'name': element}
+        fargs = {'prefix': prefix, 'type': typename}
         t = '''#include <stdio.h>
         {prefix}
         int main(int argc, char **argv) {{
-            {name} something;
+            {type} something;
         }}'''
         if not self.compiles(t.format(**fargs), env, extra_args, dependencies):
             return -1
-        t = '''#include <stdio.h>
-        {prefix}
-        int temparray[{size}-sizeof({name})];'''
-        return self._bisect_compiles(t, fargs, env, extra_args, dependencies)
+        return self.cross_compute_int('sizeof(%s)' % typename, 1, 128, None, prefix, env, extra_args, dependencies)
```

Comment:
With this change, you have reduced the max size that can be detected on a cross compiled system from 1024 to 128.  This breaks the cross compile of systemd (at least on Gentoo).  See https://bugs.gentoo.org/654026

Reply:
Wait.  Now I see this is cross_sizeof with a limit of 128, but there is also cross_alignment with a limit of 1024.  So I am not sure if a previous limit of 1024 has been reduced.  In any case the referenced Gentoo bug is calling cross_sizeof with a structure whose size is 256 (same size on build and host systems).  So the 128 byte limit of cross_sizeof seems unnecessarily low, especially since the number of guesses required for a given max value N is log_2(N).

Reply:
I now see the latest code in master has changed.  Not sure if the systemd build is using the older verison or if it is passing in a guess of 128 to the new code.

Reply:
Right, I shouldn't have reduced it, so please fix it and submit a pull request. I'm pretty sure the Meson team would accept it based on the concrete case you presented.

Reply:
Turns out it has already been fixed (see my previous comment).  I actually discovered it had been fixed when I was in the process of creating a branch for my fix ;)